### ISSUE TITLE
Performance regression due to the new scheduler

### ISSUE THREAD

**Issue by tatetian** [asterinas/asterinas#1676]
The recently-merged new task schedulers ([commit 878f3f37600ebacecd0e31c16ae04e8ba83eeec4](https://github.com/asterinas/asterinas/commit/878f3f37600ebacecd0e31c16ae04e8ba83eeec4)) has caused the performance results of LMbench to degrade significantly. A partial list of the affected benchmarks is below:

* [Process] The cost of context switching
* [Pipes] The cost of write+read (1B)
* [Pipes] The bandwidth
* [FIFO] The cost of write+read (1B)
* [Unix sockets] The latency of write+read
* [Unix sockets] The bandwidth
* [TCP sockets] The bandwidth (localhost, 4KB message)
* [TCP sockets] The bandwidth (localhost, 64KB message)
* [TCP sockets] The latency of write+read
* [TCP sockets] The latency of connect

The run of benchmark CI is [this](https://github.com/asterinas/asterinas/actions/runs/12151522638), which reported at least 14 performance regression alarms.

<img width="736" alt="image" src="https://github.com/user-attachments/assets/12030e5d-974a-4eb5-b26a-17844c34dc89">

Due to performance fluctuation, our benchmark CI regularly reports false alarms on performance regression. Normally, around 5 alarms are reported. But the CI benchmark conducted right after the new schedulers is merged raised far more alarms. So there are definitely some performance problems with the new scheduler.

**Comment by js2xxx** [comment]
I found most of the regression comes from the lock of the scheduling policy during the whole process of updating/enqueuing threads, which is indicated by the flame graph of the `process_lat_ctx` benchmark. So I opened a PR to fix it: #1678.

![截图 2024-12-04 23-33-43](https://github.com/user-attachments/assets/895f0897-60a8-40dc-949a-19e6bb3f14d9)

**Comment by tangruize** [comment]
The new scheduler also causes `cpu_affinity` to fail.

Test patch:

```diff
diff --git a/kernel/src/lib.rs b/kernel/src/lib.rs
index 3ea32c83..0acf522e 100644
--- a/kernel/src/lib.rs
+++ b/kernel/src/lib.rs
@@ -35,7 +35,7 @@ use core::sync::atomic::Ordering;
 use ostd::{
     arch::qemu::{exit_qemu, QemuExitCode},
     boot,
-    cpu::PinCurrentCpu,
+    cpu::{num_cpus, CpuId, CpuSet, PinCurrentCpu},
 };
 use process::Process;

@@ -133,10 +133,12 @@ fn init_thread() {
     net::lazy_init();
     fs::lazy_init();
     ipc::init();
+    let affinity = CpuSet::try_from(CpuId::try_from(num_cpus() - 1).unwrap()).unwrap();
     // driver::pci::virtio::block::block_device_test();
     let thread = ThreadOptions::new(|| {
         println!("[kernel] Hello world from kernel!");
     })
+    .cpu_affinity(affinity)
     .spawn();
     thread.join();

```

Result: panic with error message:

```txt
[     9.114] ERROR: Uncaught panic:
        called `Option::unwrap()` on a `None` value
        at /root/asterinas/kernel/src/thread/mod.rs:77
        on CPU 1 by thread Some(Thread { task: (Weak), data: Any { .. }, status: AtomicThreadStatus(1), priority: AtomicPriority(139), cpu_affinity: AtomicCpuSet { bits: [2] }, sched_attr: SchedAttr { policy: UnsafeCell { .. }, real_time: RealTimeAttr { prio: 99, time_slice: 44876580 }, fair: FairAttr { weight: 14, vruntime: 17950637 } } })

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panic.rs:107
(  2) /root/asterinas/kernel/src/thread/oops.rs:124
(  3) ??:?
(  4) 8jvq36gkk9o1wwgpfnq112dj2:?
(  5) ??:?
(  6) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:222
(  7) ??:?
(  8) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:971
(  9) /root/asterinas/kernel/src/sched/sched_class/mod.rs:314
( 10) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1446
( 11) /root/asterinas/kernel/src/sched/sched_class/mod.rs:327
( 12) /root/asterinas/ostd/src/task/scheduler/mod.rs:219
( 13) /root/asterinas/ostd/src/task/scheduler/mod.rs:238
( 14) /root/asterinas/kernel/src/sched/sched_class/mod.rs:241
( 15) /root/asterinas/ostd/src/task/scheduler/mod.rs:237
( 16) /root/asterinas/ostd/src/task/scheduler/mod.rs:225
( 17) /root/asterinas/ostd/src/task/mod.rs:88
( 18) /root/asterinas/kernel/src/thread/mod.rs:156
( 19) /root/asterinas/kernel/src/lib.rs:114
( 20) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 21) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 22) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1972
( 23) /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/unwinding-0.2.4/src/panicking.rs:56
( 24) 8xltkxihms6m7w2agunn1aynu:?
( 25) /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/unwinding-0.2.4/src/panicking.rs:42
( 26) /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/unwinding-0.2.4/src/panic.rs:87
( 27) /root/asterinas/kernel/src/thread/oops.rs:54
( 28) /root/asterinas/kernel/src/thread/kernel_thread.rs:53
( 29) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 30) /root/.rustup/toolchains/nightly-2024-11-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1972
( 31) /root/asterinas/ostd/src/task/mod.rs:198
```

**Comment by jellllly420** [comment]
Just to update, performance of the current (6de5be1ecc345edb5dc9d18b27cb1ca2b179b2aa) CFS implementation is comparable with that of the old scheduler on lmbench now. [Here](https://lizard-cattle-287.notion.site/17ab19f0b33680179335f7f3bd253076?v=055fb498c75f45a48139a2d5cfdb9e7e&pvs=4) are the results.

I think it's time to re-enable the CFS now (or maybe after #1790).

**Comment by tatetian** [comment]
> Just to update, performance of the current ([6de5be1](https://github.com/asterinas/asterinas/commit/6de5be1ecc345edb5dc9d18b27cb1ca2b179b2aa)) CFS implementation is comparable with that of the old scheduler on lmbench now. [Here](https://lizard-cattle-287.notion.site/17ab19f0b33680179335f7f3bd253076?v=055fb498c75f45a48139a2d5cfdb9e7e&pvs=4) are the results.

Thanks for posting the performance results. I am confused with the numbers. I have two questions:
1. They seems to be normalized. What is the baseline?
2. The higher the values, the better the results. Am I correct?

> I think it's time to re-enable the CFS now (or maybe after [#1790](https://github.com/asterinas/asterinas/pull/1790)).

Sure. If the new scheduler catches up in terms of performance, then we can switch to the new scheduler.

**Comment by jellllly420** [comment]
@tatetian Its's normalized against Linux just like we do in [asterinas's benchmark page](https://asterinas.github.io/benchmark/). I'm showing a `Performance of Asterinas / Performance of Linux` here so yes it's better to have a higher value.

**Comment by tatetian** [comment]
> [@tatetian](https://github.com/tatetian) Its's normalized against Linux just like we do in [asterinas's benchmark page](https://asterinas.github.io/benchmark/). I'm showing a `Performance of Asterinas / Performance of Linux` here so yes it's better to have a higher value.

Thanks for the clarification. The results are encouraging. What about hackbench and schbench?

**Comment by jellllly420** [comment]
The performance of the two schedulers are close on hackbench and schbench though neither is as scalable as that of Linux.
- hackbench (SMP = 1)
  | Scheduler        | Turnaround Time |
  |------------------|-----------------|
  | FIFO (Asterinas) | 1.752           |
  | CFS (Asterinas)  | 2.035                |
  | CFS (Linux)      | 1.332           |
- schbench (SMP = 1)
  | Scheduler        | P99 Wakeup Latency | P99 Request Latency | P50 RPS |
  |------------------|--------------------|---------------------|---------|
  | FIFO (Asterinas) | 1                  | 5128                | 232     |
  | CFS (Asterinas)  | 3                   | 5176                    | 231        |
  | CFS (Linux)      | 2                  | 6456                | 159     |
- schbench(SMP = 8)
  | Scheduler        | P99 Wakeup Latency | P99 Request Latency | P50 RPS |
  |------------------|--------------------|---------------------|---------|
  | FIFO (Asterinas) | 274944             | 389632              | 50      |
  | CFS (Asterinas)  | 255744                   | 596992                    | 55        |
  | CFS (Linux)      | 3964               | 39104               | 304     |

**Comment by arthurpaulino** [comment]
Are there current efforts on this front? These numbers look a bit scary at first.

**Comment by tatetian** [comment]
The performance degradation was largely addressed thanks to a series of improvements.
